### PR TITLE
Move QgsRasterLayerProperties class  to the gui API

### PR DESCRIPTION
**Pull request by vcloarec**
## Description
Motivation :
Move the `QgsRasterLayerProperties` class to the GUI API.
So, this class could be used using the QGIS API, notably for standalone applications.

The instance of this class was dependent, since QGIS 3.8, from the `QgisApp` instance, when calling the `askUserForDatumTransform()`. As this method, defines a title and only call `QgsDatumTransformDialog::run()` method from the gui API, this PR purpose to :
- move the files of the `QgsRasterLayerProperties` class ant the `QgsMapLayerStyleGuiUtils` class to the gui sources
- make those classes public in  the GUI API (GUI_EXPORT)
- change some `#include` and CMakeList.txt files in accordance

## Checklist
- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [x] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

### CODE REVIEW

**Comment by PeterPetrik** [comment]
what is the status of this PR? shall we merge it rather sooner than later?

**Comment by PeterPetrik** [comment]
@nyalldawson are you happy to merge it?

**Comment by vcloarec** [comment]
merging will be appreciated!

File: python/gui/auto_generated/qgsmaplayerstyleguiutils.sip.in [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ *                                                                      *
+ * Do not edit manually ! Edit header and run scripts/sipify.pl again   *
+ ************************************************************************/
+
+
+
+
+
+
+class QgsMapLayerStyleGuiUtils : QObject
```
Comment (nyalldawson) [inline python/gui/auto_generated/qgsmaplayerstyleguiutils.sip.in]: Does this need to be exposed to python?

File: python/gui/auto_generated/raster/qgsrasterlayerproperties.sip.in [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+Constructor
+
+:param lyr: Map layer for which properties will be displayed
+:param canvas: the QgsMapCanvas instance
+:param parent: the parent of this widget
+:param fl: windows flag
+%End
+
+
+  public slots:
```
Comment (nyalldawson) [inline python/gui/auto_generated/raster/qgsrasterlayerproperties.sip.in]: This should be made private -- we don't want to expose this

File: python/gui/auto_generated/raster/qgsrasterlayerproperties.sip.in [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+.. versionadded:: 3.10
+%End
+
+%TypeHeaderCode
+#include "qgsrasterlayerproperties.h"
+%End
+  public:
+
+    enum StyleType
```
Comment (nyalldawson) [inline python/gui/auto_generated/raster/qgsrasterlayerproperties.sip.in]: Hide this from the python bindings too, by wrapping the definition in the .h file with

```
#ifndef SIPRUN

...
#endif
```

File: src/gui/raster/qgsrasterlayerproperties.cpp [hunk 4]
```diff
@@ -1272,7 +1273,7 @@ void QgsRasterLayerProperties::pbnAddValuesManually_clicked()
 
 void QgsRasterLayerProperties::mCrsSelector_crsChanged( const QgsCoordinateReferenceSystem &crs )
 {
-  QgisApp::instance()->askUserForDatumTransform( crs, QgsProject::instance()->crs(), mRasterLayer );
+  QgsDatumTransformDialog::run( crs, QgsProject::instance()->crs(), this, mMapCanvas, tr( "Select Transformation for the raster layer" ) );
```
Comment (nyalldawson) [inline src/gui/raster/qgsrasterlayerproperties.cpp]:
```suggestion
  QgsDatumTransformDialog::run( crs, QgsProject::instance()->crs(), this, mMapCanvas, tr( "Select Transformation" ) );
```

File: python/gui/auto_generated/raster/qgsrasterlayerproperties.sip.in (hunk 2 above)
Comment (vcloarec) [reply python/gui/auto_generated/raster/qgsrasterlayerproperties.sip.in]: It seems SIP doesn't like when I make this slot private, I am investigating why... I've got a compilation error from a file related to SIP (sip_guipart0.cpp) that contains the method `sipQgsRasterLayerProperties::optionsStackedWidget_CurrentChanged(int a0)`.
The error message says that `QgsRasterLayerProperties::optionsStackedWidget_CurrentChanged(int)` is private. I am not sure to understand why this method is used elsewhere than in `QgsRasterLayerProperties`...

**Comment by nyalldawson** [comment]
You'll need to update the sip bindings after making it private - run scripts/sipify_all.sh

File: python/gui/auto_generated/raster/qgsrasterlayerproperties.sip.in (hunk 2 above)
Comment (vcloarec) [reply python/gui/auto_generated/raster/qgsrasterlayerproperties.sip.in]: Not expose it for python or for python and c++ API ? When I make this slot private (c++), I've got a compile error from a file related to sip even if run the sipify_all.sh script... I have to investigate why but I have difficulties to understand this sip issue...

**Comment by vcloarec** [comment]
> run scripts/sipify_all.sh

Sorry, I didn't have time to look at this. I've tried to "sipify_all", but compile error still here.
If I make it protected, no problem. But I suppose it is not satisfying...

**Comment by vcloarec** [comment]
@nyalldawson 
the enum isn't exposed to SIP anymore. The slot method is now protected and not expose to SIP. If this method is private, QGIS doesn't compile because a SIP issue.

**Comment by vcloarec** [comment]
ok, as a newbie with SIP, I was quite long to understand all the whys and the wherefores.

Now, only the constructor of the `QgsRasterLayerProperties` class is exposed to python. Maybe it could be useful to open this widget from python code.

If, this PR seems good to you, merging will be very appreciate, and I am open for discussion.

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Perfect -- thanks for seeing this one through!

**Pull request merged**